flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -769,9 +769,10 @@ public OperatorStateBackend createOperatorStateBackend(
 		cancelables.registerClosable(keyedStateBackend);
 
 		// restore if we have some old state
-		if (null != restoreStateHandles && null != restoreStateHandles.getManagedKeyedState()) {
-			keyedStateBackend.restore(restoreStateHandles.getManagedKeyedState());
-		}
+		Collection<KeyedStateHandle> restoreKeyedStateHandles =
+			restoreStateHandles == null ? null : restoreStateHandles.getManagedKeyedState();
+
+		keyedStateBackend.restore(restoreKeyedStateHandles);
```

Comment:
I attempted to put the restore state in the constructor as we discussed. But it turns out impossible. 

All state backends should be registered in the task so that the backends can be closed when the task is canceled.  If we put the restoring in the constructor of the backends, the construction of the backends may be blocked (e.g., due to the access to HDFS). Since the construction is not completed yet, the backend will not be registered and hence will not be closed.